ftl/benchmark/benchmark.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2017 Google Inc. All rights reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import subprocess
+import datetime
+import time
+import os
+import logging
+from google.cloud import bigquery
+
+
+class Benchmark():
+    def __init__(self, args, project, dataset, table, runtime):
+        self._base = args.base
+        self._name = args.name
+        self._directory = args.directory
+        self._iterations = args.iterations
+        self._description = args.description
+        self._project = project
+        self._dataset = dataset
+        self._table = table
+        self._runtime = runtime
+
+    def _record_build_times_to_bigquery(self, build_times):
+        current_date = datetime.datetime.now()
+        logging.info('Retrieving bigquery client')
+        client = bigquery.Client(project=self._project)
+
+        dataset_ref = client.dataset(self._dataset)
+        table_ref = dataset_ref.table(self._table)
+        table = client.get_table(table_ref)
+
+        full_name = "{0}:{1}.{2}".format(self._project,
+                                         self._dataset,
+                                         self._table)
+
+        logging.info("Adding build time data to {0}".format(full_name))
+        rows = [(current_date, self._description, bt[0], bt[1]) for
+                bt in build_times]
+        client.create_rows(table, rows)
+        logging.info("Finished adding build times to {0}".format(full_name))
+
+    def run_benchmarks(self):
+        logging.getLogger().setLevel("NOTSET")
+        logging.basicConfig(
+            format='%(asctime)s.%(msecs)03d %(levelname)-8s %(message)s',
+            datefmt='%Y-%m-%d,%H:%M:%S')
+        build_times = []
+        logging.info('Beginning building {0} images'.format(self._runtime))
+        for _ in range(self._iterations):
+            start_time = time.time()
+
+            # For the binary
+            builder_path = 'ftl/{0}_builder.par'.format(self._runtime)
+
+            # For container builder
+            if not os.path.isfile(builder_path):
+                builder_path = 'bazel-bin/ftl/{0}_builder.par'.format(
```

Comment:
There is an underlying assumption here that .par file should be present. Its probably an output of the bazel rule.
There is a possibility even this file does not exists.
Can we put L61 to L83 inside a try Except OSError?
Also raise the exception with an appropriate message explaining the assumptions.

Reply:
piggybacking on this, currently if the ftl.par file exists in the original repo, the build will fail.  a bazel clean is required to be run before launching the ftl_benchmark which shouldn't be the casel.  we should try to address that as well

Reply:
yah, I'm not exactly sure how to because I don't think there's any way of only running the build step if there isn't a par file